### PR TITLE
authorPatchCollector.py: Add GSOC patch collector

### PR DESCRIPTION
**Pull request by srivama**
Closes https://github.com/coala/meta/issues/124

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [ ] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [ ] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `cobot mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

**Comment by li-boxuan** [comment]
Migrated from https://github.com/coala/meta/pull/126

File: tools/authorPatchCollector.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+            begin = hash
+            range_count = range_count + 1
+
+        else:
+            subdir_msg = ('\nPlease enter a subdirectory name for this patch'
+                          '(leave this empty\nto leave this patch out, use ./ '
+                          'for the root directory):')
+            print('\n')
+
+            if end == "":
```
Comment (jayvdb) [inline tools/authorPatchCollector.py]: single quotes (still a few of them throughout the file )

File: tools/authorPatchCollector.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+
+def main():
+    help_message = """This is a simple helper script that allows collecting all
+        patches from one author in a git repository. It will detect consequent
+        ranges of patches and ask for each range for a subdirectory so you can
+        distinguish the patch series."""
+
+    parser = argparse.ArgumentParser(description=help_message)
+    parser.add_argument('author_names',
```
Comment (jayvdb) [inline tools/authorPatchCollector.py]: allow multiple `--author` , like `git` does

File: tools/authorPatchCollector.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+                        help="Name of contributors separated by ','")
+    parser.add_argument('output_root_dir', help='Output directory path')
+    parser.add_argument('input_dir_paths',
```
Comment (jayvdb) [inline tools/authorPatchCollector.py]: what about defaulting to reading the subdirectories of the current directory, checking each has a `.git` .

that is how most people would arrange all of the coala repos they have contributed to.

and/or, use multiple `--input-dir`

File: tools/authorPatchCollector.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+                print(f'Adding one patch to {temp} existing:')
+                subprocess.Popen(
+                    f'git -C {dir_path} log {hash}^..{hash}', shell=True) \
+                    .wait()
+                print(subdir_msg)
+                subdir = input()
+
+                if subdir != '':
+                    subprocess.Popen(
+                        f'mkdir -p {output}/{subdir}', shell=True).wait()
```
Comment (jayvdb) [inline tools/authorPatchCollector.py]: `os.mkdir`

File: tools/authorPatchCollector.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    for author in authors:
+        author_list += f'--author={author} '
+
+    get_hashes_cmd = (f'git -C {dir_path} log --pretty=format:"%h" '
+                      f'{author_list}')
+    hashes = subprocess.check_output(get_hashes_cmd, shell=True)
+    hashes = hashes.decode(encoding='utf-8')
+    hashes = hashes.split('\n')
+
+    for hash in hashes:
```
Comment (jayvdb) [inline tools/authorPatchCollector.py]: hash is a reserved word

use commit_hash ?

File: tools/authorPatchCollector.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+                        f'mkdir -p {output}/{subdir}', shell=True).wait()
+                    subprocess.Popen(
+                        f'git -C {dir_path} format-patch {begin}^^..{end} '
+                        f'-o{output}/{subdir}', shell=True).wait()
+                    patch_count = patch_count + range_count
+
+                else:
+                    print('Omitting patch series...')
+
+                end = ""
```
Comment (jayvdb) [inline tools/authorPatchCollector.py]: this use of `end` and `start` feels like they are booleans, or loop variables, and shouldnt be strings.

File: tools/authorPatchCollector.py (hunk 2 above)
Comment (srivama) [reply tools/authorPatchCollector.py]: Is `python patch.py -a author1 author2 author3 ... -o output -d dir1 dir2 dir3 ...` format fine?

Comment (srivama) [reply tools/authorPatchCollector.py]: Fixed, did it like git ;)

File: tools/authorPatchCollector.py (hunk 5 above)
Comment (srivama) [reply tools/authorPatchCollector.py]: Sure!

File: tools/authorPatchCollector.py (hunk 6 above)
Comment (srivama) [reply tools/authorPatchCollector.py]: They're used for preparing the patches, line 64 and 75.

Comment (jayvdb) [reply tools/authorPatchCollector.py]: ah I see, then again give them meaningful names , at least `foo_commit` or `foo_sha`

Comment (srivama) [reply tools/authorPatchCollector.py]: `begin_hash` and `end_hash` would suffice, I guess?

File: tools/authorPatchCollector.py (hunk 4 above)
Comment (srivama) [reply tools/authorPatchCollector.py]: Done.

File: tools/authorPatchCollector.py (hunk 3 above)
Comment (srivama) [reply tools/authorPatchCollector.py]: Added, thanks :)

**Comment by jayvdb** [comment]
ack https://github.com/coala/projects/pull/668/commits/5e05aa1584039aaf77fe1901f14dd45e329b494a

**Comment by jayvdb** [comment]
@gitmate-bot ff

**Pull request merged**